### PR TITLE
Paperwork 1.2.4

### PR DESCRIPTION
**Pull request by symphorien**
###### Motivation for this change
update (most of these updates could be made standalone. tell me if you want to split.)

###### Things done

The new home for paperwork is the gnome gitlab instance, but they have an unusual namespacing. Instead of the "traditional" owner/repo url, there are three levels: group/owner/repo: https://gitlab.gnome.org/World/OpenPaperwork/paperwork

I have adapted `fetchFromGitLab` to take an optional `group` argument.

I have checked that I can still scan/ocr succesfully with paperwork.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`) (just paperwork)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @aszlig as maintainer.

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -288,13 +288,13 @@ with pkgs;
...
+    owner, repo, rev, domain ? "gitlab.com", name ? "source", group ? null,
     ... # For hash agility
   }@args: fetchzip ({
     inherit name;
-    url = "https://${domain}/api/v4/projects/${owner}%2F${repo}/repository/archive.tar.gz?sha=${rev}";
-    meta.homepage = "https://${domain}/${owner}/${repo}/";
-  } // removeAttrs args [ "domain" "owner" "repo" "rev" ]) // { inherit rev; };
+    url = "https://${domain}/api/v4/projects/${lib.optionalString (group != null) group+"%2F"}${owner}%2F${repo}/repository/archive.tar.gz?sha=${rev}";
+    meta.homepage = "https://${domain}/${lib.optionalString (group != null) group+"/"}${owner}/${repo}/";
+  } // removeAttrs args [ "domain" "owner" "group" "repo" "rev" ]) // { inherit rev; };
```
Comment (aszlig) [inline pkgs/top-level/all-packages.nix]: Can you please split this into a separate commit?

**Review by aszlig — APPROVED** [review]
Looks good to me, thanks for the cleanups.

Given that the backend and frontend are now in one repositories it would probably make more sense to split out the source into a separate Nix expression file (eg. `source.nix`) and directly use `src = "${source}/paperwork-backend";` so that we don't need to set the source root anymore.

**Comment by aszlig** [comment]
Tested this also on NixOS, so I ticked off the checkboxes accordingly.

@symphorien: Can you please ping me again after you've split off the `fetchFromGitLab`?

**Comment by symphorien** [comment]
@aszlig I just split the first commit as requested.
Regarding paperwork-src: the problem is that it does not belong either to nixpkgs top-level nor to pythonPackages. So I somewhat prefer the current situation. But if you prefer I can move the source to its own attribute.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (ryantm) [reply pkgs/top-level/all-packages.nix]: This breaks `fetchFromGitLab`.

Before:
`https://gitlab.com/api/v4/projects/psmisc%2Fpsmisc/repository/archive.tar.gz?sha=v23.2`
After:
`https://gitlab.com/api/v4/projects/%2Fpsmisc%2Fpsmisc/repository/archive.tar.gz?sha=v23.2`

Comment (ryantm) [reply pkgs/top-level/all-packages.nix]: @azlig I'd like to revert this PR until this is resolved, okay?

Comment (aszlig) [reply pkgs/top-level/all-packages.nix]: @ryantm: Ah, this is just a grouping issue, going to fix that instead.

Comment (ryantm) [reply pkgs/top-level/all-packages.nix]: More context https://github.com/NixOS/nixpkgs/pull/46977#issuecomment-424722471

Comment (aszlig) [reply pkgs/top-level/all-packages.nix]: Fixed in 593062ddc2104cf91b1cf9cc0e4f4b1437f59f6b.
